osu.Game/Overlays/FirstRunSetupOverlay.cs

```diff
@@ -0,0 +1,397 @@
+                // Either this dialog would be converted to its own screen, or at very least be "hosted" by a screen pushed to the main menu.
+                // Alternatively, another method of disabling notifications could be added to `INotificationOverlay`.
+                if (menu != null)
+                {
+                    overlayActivationMode = menu.OverlayActivationMode.GetBoundCopy();
+                    overlayActivationMode.Value = OverlayActivation.UserTriggered;
+                }
+
+                base.Show();
+            }, new[] { typeof(MainMenu) });
+        }
+
+        protected override void PopIn()
+        {
+            base.PopIn();
+
+            this.ScaleTo(scale_when_hidden)
+                .ScaleTo(1, 400, Easing.OutElasticHalf);
+
+            this.FadeIn(400, Easing.OutQuint);
+
+            if (currentStepIndex == null)
+                showFirstStep();
+        }
+
+        protected override void PopOut()
+        {
+            if (overlayActivationMode != null)
+            {
+                // If this is non-null we are guaranteed to have come from the main menu.
+                overlayActivationMode.Value = OverlayActivation.All;
+                overlayActivationMode = null;
+            }
+
+            if (currentStepIndex != null)
+            {
+                notificationOverlay.Post(new SimpleNotification
+                {
+                    Text = FirstRunSetupOverlayStrings.ClickToResumeFirstRunSetupAtAnyPoint,
+                    Icon = FontAwesome.Solid.Horse,
+                    Activated = () =>
+                    {
+                        Show();
+                        return true;
+                    },
+                });
+            }
+            else
+            {
+                stack?.FadeOut(100)
+                     .Expire();
+            }
+
+            base.PopOut();
+
+            this.ScaleTo(0.96f, 400, Easing.OutQuint);
+            this.FadeOut(200, Easing.OutQuint);
+        }
+
+        private void showFirstStep()
+        {
+            Debug.Assert(currentStepIndex == null);
+
+            stackContainer.Child = stack = new ScreenStack
+            {
+                RelativeSizeAxes = Axes.Both,
+            };
+
+            currentStepIndex = -1;
+            showNextStep();
+        }
+
+        private void showLastStep()
+        {
+            if (currentStepIndex == 0)
+                return;
+
+            Debug.Assert(stack != null);
+
+            stack.CurrentScreen.Exit();
+            currentStepIndex--;
+
+            BackButton.Enabled.Value = currentStepIndex != 0;
+
+            updateButtonText();
+        }
+
+        private void showNextStep()
+        {
+            Debug.Assert(currentStepIndex != null);
+            Debug.Assert(stack != null);
+
+            currentStepIndex++;
+
+            BackButton.Enabled.Value = currentStepIndex > 0;
```

Comment:
This is duplicated a few times in these step navigation methods. Maybe `updateButtonText()` could be genericised into `updateButtonState()` or something and this update could be folded in there?